File: server/src/models/sentimentModel.js
const tf = require('@tensorflow/tfjs-node');
const fs = require('fs');
const csv = require('csv-parser');
const natural = require('natural');
const { WordTokenizer, SentimentAnalyzer, PorterStemmer } = natural;
const tokenizer = new WordTokenizer();

// Sentiment labels
// 0: sadness
// 1: joy
// 2: love
// 3: anger
// 4: fear
// 5: surprise

class SentimentModel {
  constructor() {
    this.model = null;
    this.wordIndex = {};
    this.maxSequenceLength = 50;
    this.vocabSize = 5000;
    this.labels = ['sadness', 'joy', 'love', 'anger', 'fear', 'surprise'];
  }

  // Preprocess text for the model
  preprocessText(text) {
    // Convert to lowercase
    text = text.toLowerCase();
    // Remove special characters and numbers
    text = text.replace(/[^\w\s]/g, '');
    text = text.replace(/\d+/g, '');
    // Tokenize
    const tokens = tokenizer.tokenize(text);
    // Convert to sequence
    const sequence = [];
    for (let i = 0; i < this.maxSequenceLength; i++) {
      if (i < tokens.length) {
        const token = tokens[i];
        if (this.wordIndex[token]) {
          sequence.push(this.wordIndex[token]);
        } else {
          sequence.push(0); // OOV token
        }
      } else {
        sequence.push(0); // Padding
      }
    }
    return sequence;
  }

  // Build vocabulary from training data
  async buildVocabulary(trainingDataPath) {
    return new Promise((resolve, reject) => {
      const wordFrequency = {};
      let totalWords = 0;

      fs.createReadStream(trainingDataPath)
        .pipe(csv())
        .on('data', (row) => {
          const text = row.text;
          const tokens = tokenizer.tokenize(text.toLowerCase());
          tokens.forEach(token => {
            if (!wordFrequency[token]) {
              wordFrequency[token] = 0;
            }
            wordFrequency[token] += 1;
            totalWords += 1;
          });
        })
        .on('end', () => {
          // Sort words by frequency
          const sortedWords = Object.keys(wordFrequency).sort((a, b) => 
            wordFrequency[b] - wordFrequency[a]
          );
          
          // Create word index (top N words)
          this.wordIndex = {};
          // Reserve 0 for padding and 1 for OOV
          this.wordIndex['<PAD>'] = 0;
          this.wordIndex['<OOV>'] = 1;
          
          // Add top words to vocabulary
          const topWords = sortedWords.slice(0, this.vocabSize - 2);
          topWords.forEach((word, index) => {
            this.wordIndex[word] = index + 2;
          });
          
          console.log(`Vocabulary built with ${Object.keys(this.wordIndex).length} words`);
          resolve();
        })
        .on('error', (error) => {
          reject(error);
        });
    });
  }

  // Load data from CSV file
  async loadData(filePath) {
    return new Promise((resolve, reject) => {
      const data = [];
      const labels = [];

      fs.createReadStream(filePath)
        .pipe(csv())
        .on('data', (row) => {
          const text = row.text;
          const label = parseInt(row.label);
          
          // Preprocess and convert to sequence
          const sequence = this.preprocessText(text);
          data.push(sequence);
          
          // One-hot encode label
          const oneHotLabel = Array(6).fill(0);
          oneHotLabel[label] = 1;
          labels.push(oneHotLabel);
        })
        .on('end', () => {
          // Convert to tensors
          const xTensor = tf.tensor2d(data);
          const yTensor = tf.tensor2d(labels);
          
          resolve({ x: xTensor, y: yTensor });
        })
        .on('error', (error) => {
          reject(error);
        });
    });
  }

  // Create the neural network model
  createModel() {
    this.model = tf.sequential();
    
    // Embedding layer
    this.model.add(tf.layers.embedding({
      inputDim: this.vocabSize,
      outputDim: 50,
      inputLength: this.maxSequenceLength
    }));
    
    // Flatten the 3D tensor to 2D
    this.model.add(tf.layers.globalAveragePooling1d());
    
    // Hidden layer
    this.model.add(tf.layers.dense({
      units: 128,
      activation: 'relu'
    }));
    
    // Dropout to prevent overfitting
    this.model.add(tf.layers.dropout({ rate: 0.5 }));
    
    // Output layer (6 emotions)
    this.model.add(tf.layers.dense({
      units: 6,
      activation: 'softmax'
    }));
    
    // Compile model
    this.model.compile({
      optimizer: 'adam',
      loss: 'categoricalCrossentropy',
      metrics: ['accuracy']
    });
    
    return this.model;
  }

  // Train the model with the provided data
  async train(trainingDataPath, validationDataPath, epochs = 10, batchSize = 32) {
    try {
      // Build vocabulary from training data
      await this.buildVocabulary(trainingDataPath);
      
      // Create model
      this.createModel();
      
      // Load training data
      console.log('Loading training data...');
      const trainingData = await this.loadData(trainingDataPath);
      
      // Load validation data
      console.log('Loading validation data...');
      const validationData = await this.loadData(validationDataPath);
      
      // Train model
      console.log('Training model...');
      await this.model.fit(trainingData.x, trainingData.y, {
        epochs,
        batchSize,
        validationData: [validationData.x, validationData.y],
        callbacks: {
          onEpochEnd: (epoch, logs) => {
            console.log(`Epoch ${epoch + 1} - loss: ${logs.loss.toFixed(4)} - accuracy: ${logs.acc.toFixed(4)} - val_loss: ${logs.val_loss.toFixed(4)} - val_acc: ${logs.val_acc.toFixed(4)}`);
          }
        }
      });
      
      console.log('Model training complete');
      return true;
    } catch (error) {
      console.error('Error training model:', error);
      return false;
    }
  }

  // Save the model to a file
  async saveModel(path) {
    if (!this.model) {
      throw new Error('No model to save');
    }
    
    try {
      // Save the model
      await this.model.save(`file://${path}`);
      
      // Save word index
      fs.writeFileSync(`${path}/word_index.json`, JSON.stringify(this.wordIndex));
      
      console.log(`Model saved to ${path}`);
      return true;
    } catch (error) {
      console.error('Error saving model:', error);
      return false;
    }
  }

  // Load the model from a file
  async loadModel(path) {
    try {
      // Load the model
      this.model = await tf.loadLayersModel(`file://${path}/model.json`);
      
      // Load word index
      const wordIndexData = fs.readFileSync(`${path}/word_index.json`, 'utf8');
      this.wordIndex = JSON.parse(wordIndexData);
      
      console.log(`Model loaded from ${path}`);
      return true;
    } catch (error) {
      console.error('Error loading model:', error);
      return false;
    }
  }

  // Predict sentiment for a text
  async predict(text) {
    if (!this.model) {
      throw new Error('Model not loaded');
    }
    
    // Preprocess text
    const sequence = this.preprocessText(text);
    
    // Make prediction
    const input = tf.tensor2d([sequence]);
    const prediction = this.model.predict(input);
    const probabilities = await prediction.data();
    
    // Get emotion with highest probability
    let maxIndex = 0;
    let maxProb = 0;
    for (let i = 0; i < probabilities.length; i++) {
      if (probabilities[i] > maxProb) {
        maxProb = probabilities[i];
        maxIndex = i;
      }
    }
    
    // Return result
    return {
      emotion: this.labels[maxIndex],
      probability: maxProb,
      allEmotions: this.labels.map((label, index) => ({
        emotion: label,
        probability: probabilities[index]
      }))
    };
  }

  // Simple sentiment analysis using natural library as fallback
  analyzeSentimentFallback(text) {
    const analyzer = new SentimentAnalyzer('English', PorterStemmer, 'afinn');
    const result = analyzer.getSentiment(tokenizer.tokenize(text));
    
    // Map result to our emotions
    // AFINN returns a score between -5 and 5
    if (result <= -3) {
      return { emotion: 'anger', probability: 0.8 };
    } else if (result < 0) {
      return { emotion: 'sadness', probability: 0.7 };
    } else if (result === 0) {
      return { emotion: 'surprise', probability: 0.5 };
    } else if (result <= 3) {
      return { emotion: 'joy', probability: 0.7 };
    } else {
      return { emotion: 'love', probability: 0.8 };
    }
  }
}

module.exports = SentimentModel; 

File: server/src/models/sentiment_model.py
import tensorflow as tf
import numpy as np
import pandas as pd
import re
import os
import json
from sklearn.preprocessing import OneHotEncoder
from nltk.tokenize import word_tokenize
import nltk

# Download necessary NLTK data
nltk.download('punkt', quiet=True)

class SentimentModel:
    def __init__(self):
        self.model = None
        self.word_index = {}
        self.max_sequence_length = 50
        self.vocab_size = 5000
        self.labels = ['sadness', 'joy', 'love', 'anger', 'fear', 'surprise']
        
    def preprocess_text(self, text):
        """Preprocess text for the model."""
        # Convert to lowercase
        text = text.lower()
        # Remove special characters and numbers
        text = re.sub(r'[^\w\s]', '', text)
        text = re.sub(r'\d+', '', text)
        # Tokenize
        tokens = word_tokenize(text)
        # Convert to sequence
        sequence = []
        for i in range(self.max_sequence_length):
            if i < len(tokens):
                token = tokens[i]
                if token in self.word_index:
                    sequence.append(self.word_index[token])
                else:
                    sequence.append(0)  # OOV token
            else:
                sequence.append(0)  # Padding
        return sequence
    
    def build_vocabulary(self, training_data_path):
        """Build vocabulary from training data."""
        print("Building vocabulary...")
        word_frequency = {}
        total_words = 0
        
        # Read CSV file
        df = pd.read_csv(training_data_path)
        
        # Process each text
        for text in df['text']:
            tokens = word_tokenize(text.lower())
            for token in tokens:
                if token not in word_frequency:
                    word_frequency[token] = 0
                word_frequency[token] += 1
                total_words += 1
        
        # Sort words by frequency
        sorted_words = sorted(word_frequency.keys(), 
                             key=lambda x: word_frequency[x], 
                             reverse=True)
        
        # Create word index (top N words)
        self.word_index = {}
        # Reserve 0 for padding and 1 for OOV
        self.word_index['<PAD>'] = 0
        self.word_index['<OOV>'] = 1
        
        # Add top words to vocabulary
        top_words = sorted_words[:self.vocab_size - 2]
        for index, word in enumerate(top_words):
            self.word_index[word] = index + 2
        
        print(f"Vocabulary built with {len(self.word_index)} words")
    
    def load_data(self, file_path):
        """Load data from CSV file."""
        print(f"Loading data from {file_path}...")
        df = pd.read_csv(file_path)
        
        # Preprocess texts
        sequences = []
        for text in df['text']:
            sequence = self.preprocess_text(text)
            sequences.append(sequence)
        
        # Convert to numpy arrays
        X = np.array(sequences)
        
        # One-hot encode labels
        encoder = OneHotEncoder(sparse=False, categories=[range(6)])
        y = encoder.fit_transform(df['label'].values.reshape(-1, 1))
        
        return X, y
    
    def create_model(self):
        """Create the neural network model."""
        model = tf.keras.Sequential([
            # Embedding layer
            tf.keras.layers.Embedding(
                input_dim=self.vocab_size,
                output_dim=50,
                input_length=self.max_sequence_length
            ),
            
            # Global average pooling
            tf.keras.layers.GlobalAveragePooling1D(),
            
            # Hidden layer
            tf.keras.layers.Dense(128, activation='relu'),
            
            # Dropout to prevent overfitting
            tf.keras.layers.Dropout(0.5),
            
            # Output layer (6 emotions)
            tf.keras.layers.Dense(6, activation='softmax')
        ])
        
        # Compile model
        model.compile(
            optimizer='adam',
            loss='categorical_crossentropy',
            metrics=['accuracy']
        )
        
        self.model = model
        return model
    
    def train(self, training_data_path, validation_data_path, epochs=10, batch_size=32):
        """Train the model with the provided data."""
        try:
            # Build vocabulary from training data
            self.build_vocabulary(training_data_path)
            
            # Create model
            self.create_model()
            
            # Load training data
            X_train, y_train = self.load_data(training_data_path)
            
            # Load validation data
            X_val, y_val = self.load_data(validation_data_path)
            
            # Train model
            print("Training model...")
            history = self.model.fit(
                X_train, y_train,
                epochs=epochs,
                batch_size=batch_size,
                validation_data=(X_val, y_val),
                verbose=1
            )
            
            print("Model training complete")
            return True, history
        except Exception as e:
            print(f"Error training model: {e}")
            return False, None
    
    def save_model(self, path):
        """Save the model to a file."""
        if self.model is None:
            raise ValueError("No model to save")
        
        try:
            # Create directory if it doesn't exist
            os.makedirs(path, exist_ok=True)
            
            # Save the model
            self.model.save(os.path.join(path, 'model.h5'))
            
            # Save word index
            with open(os.path.join(path, 'word_index.json'), 'w') as f:
                json.dump(self.word_index, f)
            
            print(f"Model saved to {path}")
            return True
        except Exception as e:
            print(f"Error saving model: {e}")
            return False
    
    def load_model(self, path):
        """Load the model from a file."""
        try:
            # Load the model
            self.model = tf.keras.models.load_model(os.path.join(path, 'model.h5'))
            
            # Load word index
            with open(os.path.join(path, 'word_index.json'), 'r') as f:
                self.word_index = json.load(f)
            
            print(f"Model loaded from {path}")
            return True
        except Exception as e:
            print(f"Error loading model: {e}")
            return False
    
    def predict(self, text):
        """Predict sentiment for a text."""
        if self.model is None:
            raise ValueError("Model not loaded")
        
        # Preprocess text
        sequence = self.preprocess_text(text)
        
        # Make prediction
        input_data = np.array([sequence])
        prediction = self.model.predict(input_data)
        probabilities = prediction[0]
        
        # Get emotion with highest probability
        max_index = np.argmax(probabilities)
        max_prob = probabilities[max_index]
        
        # Return result
        return {
            "emotion": self.labels[max_index],
            "probability": float(max_prob),
            "all_emotions": [
                {"emotion": label, "probability": float(prob)}
                for label, prob in zip(self.labels, probabilities)
            ]
        }
    
    def analyze_sentiment_fallback(self, text):
        """Simple sentiment analysis as fallback."""
        try:
            from nltk.sentiment.vader import SentimentIntensityAnalyzer
            nltk.download('vader_lexicon', quiet=True)
            
            # Use VADER for sentiment analysis
            sid = SentimentIntensityAnalyzer()
            scores = sid.polarity_scores(text)
            
            # Map VADER scores to our emotions
            compound = scores['compound']
            
            if compound <= -0.5:
                return {"emotion": "anger", "probability": 0.8}
            elif compound < 0:
                return {"emotion": "sadness", "probability": 0.7}
            elif compound == 0:
                return {"emotion": "surprise", "probability": 0.5}
            elif compound <= 0.5:
                return {"emotion": "joy", "probability": 0.7}
            else:
                return {"emotion": "love", "probability": 0.8}
        except Exception:
            # If VADER is not available, use a simpler approach
            # Count positive and negative words
            positive_words = ["good", "great", "happy", "joy", "love", "positive", "excellent", "wonderful", "amazing"]
            negative_words = ["bad", "sad", "unhappy", "angry", "hate", "negative", "terrible", "awful", "pain"]
            
            tokens = word_tokenize(text.lower())
            positive_count = sum(1 for word in tokens if word in positive_words)
            negative_count = sum(1 for word in tokens if word in negative_words)
            
            if negative_count > positive_count and negative_count > 2:
                return {"emotion": "anger", "probability": 0.8}
            elif negative_count > positive_count:
                return {"emotion": "sadness", "probability": 0.7}
            elif positive_count == negative_count:
                return {"emotion": "surprise", "probability": 0.5}
            elif positive_count > negative_count:
                return {"emotion": "joy", "probability": 0.7}
            else:
                return {"emotion": "love", "probability": 0.8}

# Example usage
if __name__ == "__main__":
    model = SentimentModel()
    
    # Train model
    training_path = "path/to/training.csv"
    validation_path = "path/to/validation.csv"
    test_path = "path/to/test.csv"
    
    if os.path.exists(training_path) and os.path.exists(validation_path):
        success, _ = model.train(training_path, validation_path, epochs=5)
        if success:
            model.save_model("./models/sentiment")
            
            # Test model
            test_text = "I feel really happy today!"
            result = model.predict(test_text)
            print(f"Text: {test_text}")
            print(f"Predicted emotion: {result['emotion']} ({result['probability']:.2f})")
    else:
        print("Using fallback sentiment analysis...")
        test_text = "I feel really sad today."
        result = model.analyze_sentiment_fallback(test_text)
        print(f"Text: {test_text}")
        print(f"Predicted emotion: {result['emotion']} ({result['probability']:.2f})") 